src/Mono.Android/Android.Content/ContentValues.cs

```diff
@@ -170,7 +170,7 @@ public void Put (string key, int value)
 				id_put_Ljava_lang_String_Ljava_lang_Integer_ = JNIEnv.GetMethodID (class_ref, "put", "(Ljava/lang/String;Ljava/lang/Integer;)V");
 			IntPtr jkey = JNIEnv.NewString (key);
 			try {
-				using (var val = new Java.Lang.Integer (value))
+				using (var val = Java.Lang.Integer.ValueOf (value))
```

Comment:
Using `Integer.valueOf()` may be the recommended alternative, but…

Our problem is going to be "instance sharing": `Integer.valueOf()` "yield[s] significantly better space and time performance" because it has a cache:

* https://cs.android.com/android/platform/superproject/main/+/main:libcore/ojluni/src/main/java/java/lang/Integer.java;l=1207?q=Integer.java

This presents the possibility of "accidental instance sharing".  What happens if two threads run:

```csharp
var v = Java.Lang.Integer.ValueOf(42);
```

What happens is that `Integer.valueOf(42)` will return the same Java-side instance, so `Integer.ValueOf()` will (attempt) to return the *same* managed callable wrapper instance.

So far, so fine.  But that `using` is what kills things; if two threads run:

```csharp
using (var val = Java.Lang.Integer.ValueOf(42)) {
    val.FloatValue();
}
```

then we have a possible "single world ordering" of:

 1. Thread 1: calls `Integer.ValueOf(42)`, calls `Integer.valueOf(42)`, wrapper instance created and registered.
 2. Thread 1: calls `val.FloatValue()`
 3. Thread 2: calls `Integer.ValueOf(42)`, gets wrapper instance created in (1)
 4. Thread 1: calls `val.Dispose()`
 5. Thread 2: calls `val.FloatValue()`.

Step (5) throws an `System.ObjectDisposedException`.  Oops.

You can *either* have `using (var v = new T(…))` *or* we can have `var v = T.valueOf(…)`.  We *cannot* intermix them, not safely.

(Of course, I have [some](https://github.com/xamarin/java.interop/issues/4) [ideas](https://github.com/jonpryor/java.interop/commit/a675a8796acb5968f0b7b9a2b7261b7878730dcc) to make this "more reasonable", but they have not been finished yet.)

We thus have a question: which is "worse":

 1. the extra GREFs that will happen by using `Integer.ValueOf()` without a `using` block, or
 2. the extra Java-side (and C# side!) `Integer` instances which will be created by using the constructor?

That would be a fascinating microbenchmark to see the results of.

*For now*, I think we should stick with the current `using` approach and continue using the constructors.

Reply:
The same logic applies to all the other types.